services/application.go

```diff
@@ -80,9 +80,13 @@ func (app *ChainlinkApplication) Start() error {
 
 	return multierr.Combine(
 		app.Store.Start(),
-		app.HeadTracker.Start(),
-		app.Scheduler.Start(),
 		app.JobRunner.Start(),
+		app.Scheduler.Start(),
+		// HeadTracker deliberately started after JobRunner to prevent race
+		// condition between HeadTracker#Connect->JobSubscriber initiating
+		// runs from logs and JobRunner#Start resuming runs since last shutdown.
+		// https://www.pivotaltracker.com/story/show/162230780
+		app.HeadTracker.Start(),
```

Comment:
This seems a little subtle, both of these functions appear to have parts that are blocking and parts that are in a goroutine, but they interact in complex ways.

Perhaps we could split that part out to make it more obvious? 

Maybe a separate Init and Start part?

Reply:
I agree with you, it is subtle, and the order is not explicitly tested. Here is a breakdown of how I understand it:

### Channel
It's all running through the shared `store.RunChannel`:

### Producers:
1. JobRunner#resumeRunsSinceLastShutdown (in the `main` go routine)
2. HeadTracker#backfillLogs -> JobSubscriber#dispatchLogs (in `go sub.listenToLogs(filter)` routine)

### Consumers:
1. JobRunner#demultiplexRuns (in its own go routine)

Open to ideas on how make this more obvious and explicit. I'm also cognizant of the size of the PR.

Is the idea behind separate `Init` from `Start` to help communicate that Producer 1. should happen before Producer 2? Technically, right now, I don't think anything blocks, since the Consumer `JobRunner#demultiplexRuns` starts before any of the producers. Perhaps that's in the init? Open to discuss offline in slack.

Reply:
Now that I think about it some more, I believe the awkwardness lies in `JobRunner` having to consume (`demultiplexRuns`) its own production (`resumeRunsSinceLastShutdown`) through the global channel `store.RunChannel`, even though both are internal to `JobRunner` and happen in `Start`.

Will think about this some more...